docs/guide/ethereum-provider.md

```diff
@@ -112,83 +182,394 @@ params: [
+
+::: tip
+The [`eth-rpc-errors`](https://npmjs.com/package/eth-rpc-errors) package implements all RPC errors thrown by the MetaMask provider, and can help you identify their meaning.
+:::
+
+## Using the Provider
+
+This snippet explains how to accomplish the three most common requirements for web3 sites:
+
+- Detect the Ethereum provider (`window.ethereum`)
+- Detect which Ethereum network the user is connected to
+- Get the user's Ethereum account(s)
+
+<<< @/docs/snippets/handleProvider.js
+
+## Experimental API
 
 ::: warning
-Use `ethereum.sendAsync({ method: 'eth_requestAccounts' }, callback)` instead.
+There is no guarantee that the methods and properties defined in this section will remain stable.
+Use it at your own risk.
 :::
 
-Requests to view the user's Ethereum address.
-Returns a promise of an array of hex-prefixed ethereum address strings.
+We expose some experimental, MetaMask-specific methods under the `ethereum._metamask` property.
 
-### ethereum.send(payload, callback) (DEPRECATED)
+## Experimental Methods
+
+### ethereum.\_metamask.isApproved() (TO BE REMOVED)
+
+::: danger DANGER
+This will be removed in the future.
+
+Please see the [Using the Provider section](#using-the-provider) for the recommended way of keeping track of user accounts.
+:::
+
+```typescript
+ethereum._metamask.isApproved(): Promise<boolean>;
+```
+
+This method returns a `Promise` that resolves to a `boolean` indicating if the caller has access to user accounts.
+
+### ethereum.\_metamask.isEnabled() (TO BE REMOVED)
+
+::: danger DANGER
+This will be removed in the future.
+
+Please see the [Using the Provider section](#using-the-provider) for the recommended way of keeping track of user accounts.
+:::
+
+```typescript
+ethereum._metamask.isEnabled(): boolean;
+```
+
+This method returns a `boolean` indicating if the caller has access to user accounts.
+
+### ethereum.\_metamask.isUnlocked()
+
+```typescript
+ethereum._metamask.isUnlocked(): Promise<boolean>;
+```
+
+This method returns a `Promise` that resolves to a `boolean` indicating if MetaMask is unlocked by the user.
+MetaMask must be unlocked in order to perform any operation involving user accounts.
+Note that this method does not indicate if the user has exposed any accounts to the caller.
+
+## Legacy API
 
 ::: warning
-Use `ethereum.sendAsync()` instead.
+You should **never** rely on any of these methods, properties, or events in practice.
 :::
 
-See `ethereum.sendAsync`, directly below.
+This section documents our legacy provider API.
+MetaMask only supported this API before the provider API was standardized via [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) in 2020.
+Because of this, you may find web3 sites that use this API, or other providers that implement it.
 
-### ethereum.on(eventName, callback)
+## Legacy Properties
 
-The provider supports listening for some events:
+### ethereum.autoRefreshOnNetworkChange (DEPRECATED)
```

Comment:
I don't think this should be marked as deprecated yet either. It doesn't make sense to deprecate something we're still actively encouraging people to use. We'd be giving dapp developers contradictory instructions.
```suggestion
### ethereum.autoRefreshOnNetworkChange
```

Reply:
882eb8a